source/fe/mapping_fe.cc

```diff
@@ -0,0 +1,1385 @@
+#include <deal.II/base/qprojector.h>
+#include <deal.II/base/memory_consumption.h>
+#include <deal.II/base/tensor_product_polynomials.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/grid/tria_iterator.h>
+#include <deal.II/grid/tria_boundary.h>
+#include <deal.II/dofs/dof_accessor.h>
+#include <deal.II/fe/fe_tools.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/fe_system.h>
+#include <deal.II/fe/mapping_fe.h>
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/mapping.h>
+#include <deal.II/fe/mapping_q1.h>
+#include <deal.II/base/qprojector.h>
+#include <deal.II/base/thread_management.h>
+#include <deal.II/numerics/vector_tools.h>
+
+#include <numeric>
+#include <memory>
+#include <fstream>
+
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+MappingFE<dim,spacedim,DH,VECTOR>::InternalData::InternalData (const FiniteElement<dim,spacedim> &fe,
+    const ComponentMask mask)
+  :
+  n_shape_functions (fe.dofs_per_cell),
+  mask (mask)
+{}
+
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+std::size_t
+MappingFE<dim,spacedim,DH,VECTOR>::InternalData::memory_consumption () const
+{
+  return 0;
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+MappingFE<dim,spacedim,DH,VECTOR>::MappingFE (const VECTOR  &euler_vector,
+                                              const DH      &euler_dof_handler,
+                                              const ComponentMask mask)
+  :
+  euler_vector(&euler_vector),
+  fe(&euler_dof_handler.get_fe()),
+  euler_dof_handler(&euler_dof_handler),
+  local_dofs(fe->dofs_per_cell),
+  dof_indices(fe->dofs_per_cell),
+  fe_mask(mask.size() ? mask :
+          ComponentMask(fe->get_nonzero_components(0).size(), true)),
+  fe_to_real(fe_mask.size(), numbers::invalid_unsigned_int)
+{
+
+  unsigned int size = 0;
+  for (unsigned int i=0; i<fe_mask.size(); ++i)
+    {
+      if (fe_mask[i])
+        fe_to_real[i] = size++;
+    }
+  AssertDimension(size,spacedim);
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+MappingFE<dim,spacedim,DH,VECTOR>::MappingFE (const MappingFE<dim,spacedim,DH,VECTOR> &mapping)
+  :
+  euler_vector(mapping.euler_vector),
+  fe(mapping.fe),
+  euler_dof_handler(mapping.euler_dof_handler),
+  local_dofs(mapping.local_dofs),
+  dof_indices(mapping.dof_indices),
+  fe_mask(mapping.fe_mask),
+  fe_to_real(mapping.fe_to_real)
+{}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+MappingFE<dim,spacedim,DH,VECTOR>::~MappingFE ()
+{
+  euler_dof_handler = NULL;
+  fe = NULL;
+  euler_vector = NULL;
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+void
+MappingFE<dim,spacedim,DH,VECTOR>::compute_shapes_virtual (
+  const std::vector<Point<dim> > &unit_points,
+  typename MappingFE<dim, spacedim>::InternalData &data) const
+{
+  const unsigned int n_points=unit_points.size();
+
+  if (data.shape_values.size()!=0 || data.shape_derivatives.size()!=0)
```

Comment:
Not sure if this can ever happen, but what if `shape_grad_grad` has nonzero length but the others have zero length? Then we fail to fill these fields. I find that the loop must be cheap enough that it's not worth guarding if with the `if` here.